test/spyAdapter/spyadapter.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// NOTE: TODO : Auto-generate this file for given templates
+
+package spyAdapter
+
+import (
+	"context"
+
+	"github.com/gogo/protobuf/types"
+
+	"istio.io/mixer/pkg/adapter"
+	reportTmpl "istio.io/mixer/test/template/report"
+)
+
+type (
+
+	// Adptr is a fake Adapter. It is used for controlling the Adapter's behavior as well as
+	// inspect the input values that adapter receives from Mixer
+	// nolint
+	Adptr struct {
+		AdptBehavior    AdptBehavior
+		BldrCallData    *builderCallData
+		HandlerCallData *handlerCallData
+	}
+
+	// nolint
+	AdptBehavior struct {
+		Name            string
+		BuilderBehavior BuilderBehavior
+		HandlerBhavior  HandlerBehavior
+	}
+
+	// nolint
+	HandlerBehavior struct {
+		HandleSampleReport_Error error
+		HandleSampleReport_Panic bool
+
+		Close_Error error
+		Close_Panic bool
+	}
+
+	// nolint
+	BuilderBehavior struct {
+		SetSampleReportTypes_Panic bool
+
+		SetAdapterConfig_Panic bool
+
+		Validate_Err   *adapter.ConfigErrors
+		Validate_Panic bool
+
+		Build_Err   error
+		Build_Panic bool
+	}
+
+	// nolint
+	bldr struct {
+		builderBehavior BuilderBehavior
+		handlerBehavior HandlerBehavior
+
+		builderCallData *builderCallData
+		handlerCallData *handlerCallData
+	}
+
+	// nolint
+	hndlr struct {
+		hndlrbehavior HandlerBehavior
+		hndlrCallData *handlerCallData
+	}
+
+	// nolint
+	handlerCallData struct {
+		HandleSampleReport_Instances []*reportTmpl.Instance
+		HandleSampleReport_Ctr       int
+
+		Close_Ctr int
+	}
+
+	// nolint
+	builderCallData struct {
+		// no of time called
+		SetSampleReportTypes_Ctr int
+		// input to the method
+		SetSampleReportTypes_Types map[string]*reportTmpl.Type
+
+		SetAdapterConfig_AdptCfg adapter.Config
```

Comment:
Config

Reply:
This fields if for a specific function (ideally I would like to autogenerated this complete adapter). So functionName and the variable name makes sense here IMO.